src/ui/units/ql/containers/PanePreview/PanePreview.tsx

```diff
@@ -122,11 +139,54 @@ class Preview extends React.PureComponent<PreviewProps, PreviewState> {
                     menuType={this.getMenuType()}
                     forwardedRef={this.chartKitRef}
                     workbookId={entry?.workbookId}
+                    customMenuOptions={this.getCustomMenuOptions()}
                 />
             </div>
         );
     }
 
+    getCustomMenuOptions() {
+        return {
+            [MenuItemsIds.EXPORT]: {
+                actionWrapper: this.wrapChartKitMenuSaveChartAction.bind(
+                    null,
+                    i18n('wizard', 'confirm_chart-save_message'),
+                ),
+            },
+        } as unknown as ChartProviderPropsWithRefProps['customMenuOptions'];
+    }
+
+    wrapChartKitMenuSaveChartAction = (
+        confirmText: string,
+        originalAction: MenuItemConfig['action'],
+    ) => {
+        return (...originalActionArgs: any) => {
+            const {entryCanBeSaved} = this.props;
+            return new Promise((resolve) => {
+                if (entryCanBeSaved) {
+                    const onApply = async () => {
+                        const {previewData} = this.props;
+
+                        if (!previewData) {
+                            return;
+                        }
+
+                        const preparedChartData = prepareChartDataBeforeSave(previewData);
+                        this.props.updateChartAndDoAction(preparedChartData, () =>
+                            resolve(originalAction.apply(this, originalActionArgs)),
+                        );
+                    };
+                    this.props.openDialogSaveChartConfirm({
+                        onApply,
+                        message: confirmText,
+                    });
+                } else {
+                    resolve(originalAction.apply(this, originalActionArgs));
+                }
+            });
+        };
+    };
```

Comment:
It looks very similar to the code in the Wizard. Is there any way to combine/put into a utility/common component?

Reply:
I think it's possible to put this in a separate function, but it seems to me that it will turn out to be a lot of parameters (confirmText, entryCanBeSaved, onApply, openDialogSaveChartConfirm) and the code will become unreadable. But I'm not sure